### ISSUE TITLE
MPC-cifar example is incorrect for world_size > 3

### ISSUE THREAD

**Issue by Dong-Jie-Chen** [facebookresearch/CrypTen#381]
Hi,

When I tried mpc_cifar under the example folder, it gave the correct result when world_size is 1 or 2. But the result is incorrect for world_size > 3 and the loss during inferencing explodes.

I used "python3 examples/mpc_cifar/launcher.py" for training and 
"python examples/mpc_cifar/launcher.py \
    --evaluate \
    --resume \
    --model-location "examples/mpc_cifar/model/cifar-model-checkpoint.pth.tar" \
    --batch-size 128 \
    --print-freq 1 \
    --skip-plaintext \
    --world_size 3 \
    --multiprocess"
for inferencing.

**Comment by Dong-Jie-Chen** [comment]
INFO:root:==================
INFO:root:DistributedCommunicator with rank 2
INFO:root:==================
INFO:root:==================
INFO:root:DistributedCommunicator with rank 1
INFO:root:==================
INFO:root:==================
INFO:root:DistributedCommunicator with rank 0
INFO:root:==================
INFO:torch.distributed.distributed_c10d:Added key: store_based_barrier_key:1 to store for rank: 0
INFO:torch.distributed.distributed_c10d:Added key: store_based_barrier_key:1 to store for rank: 1
INFO:torch.distributed.distributed_c10d:Added key: store_based_barrier_key:1 to store for rank: 2
INFO:torch.distributed.distributed_c10d:Rank 1: Completed store-based barrier for 3 nodes.
INFO:torch.distributed.distributed_c10d:Rank 2: Completed store-based barrier for 3 nodes.
INFO:torch.distributed.distributed_c10d:Rank 0: Completed store-based barrier for 3 nodes.
INFO:torch.distributed.distributed_c10d:Added key: store_based_barrier_key:2 to store for rank: 1
INFO:torch.distributed.distributed_c10d:Added key: store_based_barrier_key:2 to store for rank: 0
INFO:torch.distributed.distributed_c10d:Added key: store_based_barrier_key:2 to store for rank: 2
INFO:torch.distributed.distributed_c10d:Rank 2: Completed store-based barrier for 3 nodes.
INFO:torch.distributed.distributed_c10d:Rank 0: Completed store-based barrier for 3 nodes.
INFO:torch.distributed.distributed_c10d:Rank 1: Completed store-based barrier for 3 nodes.
INFO:torch.distributed.distributed_c10d:Added key: store_based_barrier_key:3 to store for rank: 1
INFO:torch.distributed.distributed_c10d:Added key: store_based_barrier_key:3 to store for rank: 2
INFO:torch.distributed.distributed_c10d:Added key: store_based_barrier_key:3 to store for rank: 0
INFO:torch.distributed.distributed_c10d:Rank 0: Completed store-based barrier for 3 nodes.
INFO:torch.distributed.distributed_c10d:Rank 1: Completed store-based barrier for 3 nodes.
INFO:torch.distributed.distributed_c10d:Rank 2: Completed store-based barrier for 3 nodes.
INFO:torch.distributed.distributed_c10d:Added key: store_based_barrier_key:4 to store for rank: 0
INFO:torch.distributed.distributed_c10d:Added key: store_based_barrier_key:4 to store for rank: 1
INFO:torch.distributed.distributed_c10d:Added key: store_based_barrier_key:4 to store for rank: 2
INFO:torch.distributed.distributed_c10d:Rank 1: Completed store-based barrier for 3 nodes.
INFO:torch.distributed.distributed_c10d:Rank 2: Completed store-based barrier for 3 nodes.
INFO:torch.distributed.distributed_c10d:Rank 0: Completed store-based barrier for 3 nodes.
INFO:root:World size = 3
INFO:root:World size = 3
INFO:root:World size = 3
/usr/local/lib/python3.7/dist-packages/crypten/__init__.py:64: RuntimeWarning: CrypTen is already initialized.
  warnings.warn("CrypTen is already initialized.", RuntimeWarning)
INFO:root:=> loading checkpoint 'examples/mpc_cifar/model/cifar-model-checkpoint.pth.tar'
INFO:root:=> loaded checkpoint 'examples/mpc_cifar/model/cifar-model-checkpoint.pth.tar' (epoch 25)
/usr/local/lib/python3.7/dist-packages/crypten/__init__.py:64: RuntimeWarning: CrypTen is already initialized.
  warnings.warn("CrypTen is already initialized.", RuntimeWarning)
/usr/local/lib/python3.7/dist-packages/crypten/__init__.py:64: RuntimeWarning: CrypTen is already initialized.
  warnings.warn("CrypTen is already initialized.", RuntimeWarning)
Downloading https://www.cs.toronto.edu/~kriz/cifar-10-python.tar.gz to /tmp/tmprpqe5ukv/cifar-10-python.tar.gz
Downloading https://www.cs.toronto.edu/~kriz/cifar-10-python.tar.gz to /tmp/tmpylp5nkhe/cifar-10-python.tar.gz
Downloading https://www.cs.toronto.edu/~kriz/cifar-10-python.tar.gz to /tmp/tmp3js_t1b2/cifar-10-python.tar.gz
170499072it [00:11, 15452401.29it/s]                   
170499072it [00:10, 15652485.95it/s]                   
170499072it [00:10, 15623168.67it/s]                   
Extracting /tmp/tmprpqe5ukv/cifar-10-python.tar.gz to /tmp/tmprpqe5ukv
Extracting /tmp/tmpylp5nkhe/cifar-10-python.tar.gz to /tmp/tmpylp5nkhe
Extracting /tmp/tmp3js_t1b2/cifar-10-python.tar.gz to /tmp/tmp3js_t1b2
Files already downloaded and verified
Files already downloaded and verified
Files already downloaded and verified
INFO:root:===== Evaluating plaintext LeNet network =====
/usr/local/lib/python3.7/dist-packages/torch/nn/functional.py:718: UserWarning: Named tensors and all their associated APIs are an experimental feature and subject to change. Please do not use them for anything important until they are released as stable. (Triggered internally at  /pytorch/c10/core/TensorImpl.h:1156.)
  return torch.max_pool2d(input, kernel_size, stride, padding, dilation, ceil_mode)
/usr/local/lib/python3.7/dist-packages/torch/nn/functional.py:718: UserWarning: Named tensors and all their associated APIs are an experimental feature and subject to change. Please do not use them for anything important until they are released as stable. (Triggered internally at  /pytorch/c10/core/TensorImpl.h:1156.)
  return torch.max_pool2d(input, kernel_size, stride, padding, dilation, ceil_mode)
INFO:root:
Test: [1/79]	Time 14.360 (14.360)	Loss 1.1341 (1.1341)	Prec@1 62.500 (62.500)   	Prec@5 95.312 (95.312)
INFO:root:
Test: [2/79]	Time 0.087 (7.223)	Loss 1.0287 (1.0814)	Prec@1 67.188 (64.844)   	Prec@5 96.875 (96.094)
/usr/local/lib/python3.7/dist-packages/torch/nn/functional.py:718: UserWarning: Named tensors and all their associated APIs are an experimental feature and subject to change. Please do not use them for anything important until they are released as stable. (Triggered internally at  /pytorch/c10/core/TensorImpl.h:1156.)
  return torch.max_pool2d(input, kernel_size, stride, padding, dilation, ceil_mode)
INFO:root:
Test: [79/79]	Time 0.011 (0.254)	Loss 1.2952 (1.0488)	Prec@1 56.250 (65.310)   	Prec@5 100.000 (96.520)
INFO:root: * Prec@1 65.310 Prec@5 96.520
INFO:root:===== Evaluating Private LeNet network =====
/usr/local/lib/python3.7/dist-packages/crypten/nn/onnx_converter.py:161: UserWarning: The given NumPy array is not writeable, and PyTorch does not support non-writeable tensors. This means you can write to the underlying (supposedly non-writeable) NumPy array using the tensor. You may want to copy the array to protect its data or make it writeable before converting it to a tensor. This type of warning will be suppressed for the rest of this program. (Triggered internally at  /pytorch/torch/csrc/utils/tensor_numpy.cpp:180.)
  param = torch.from_numpy(numpy_helper.to_array(node))
/usr/local/lib/python3.7/dist-packages/crypten/nn/onnx_converter.py:161: UserWarning: The given NumPy array is not writeable, and PyTorch does not support non-writeable tensors. This means you can write to the underlying (supposedly non-writeable) NumPy array using the tensor. You may want to copy the array to protect its data or make it writeable before converting it to a tensor. This type of warning will be suppressed for the rest of this program. (Triggered internally at  /pytorch/torch/csrc/utils/tensor_numpy.cpp:180.)
  param = torch.from_numpy(numpy_helper.to_array(node))
/usr/local/lib/python3.7/dist-packages/crypten/nn/onnx_converter.py:161: UserWarning: The given NumPy array is not writeable, and PyTorch does not support non-writeable tensors. This means you can write to the underlying (supposedly non-writeable) NumPy array using the tensor. You may want to copy the array to protect its data or make it writeable before converting it to a tensor. This type of warning will be suppressed for the rest of this program. (Triggered internally at  /pytorch/torch/csrc/utils/tensor_numpy.cpp:180.)
  param = torch.from_numpy(numpy_helper.to_array(node))
/usr/local/lib/python3.7/dist-packages/torch/_tensor.py:575: UserWarning: floor_divide is deprecated, and will be removed in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values.
To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor'). (Triggered internally at  /pytorch/aten/src/ATen/native/BinaryOps.cpp:467.)
  return torch.floor_divide(self, other)
/usr/local/lib/python3.7/dist-packages/torch/_tensor.py:575: UserWarning: floor_divide is deprecated, and will be removed in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values.
To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor'). (Triggered internally at  /pytorch/aten/src/ATen/native/BinaryOps.cpp:467.)
  return torch.floor_divide(self, other)
INFO:root:
Test: [1/79]	Time 82.535 (82.535)	Loss 5110284288.0000 (5110284288.0000)	Prec@1 10.938 (10.938)   	Prec@5 52.344 (52.344)
/usr/local/lib/python3.7/dist-packages/torch/_tensor.py:575: UserWarning: floor_divide is deprecated, and will be removed in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values.
To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor'). (Triggered internally at  /pytorch/aten/src/ATen/native/BinaryOps.cpp:467.)
  return torch.floor_divide(self, other)
INFO:root:
Test: [2/79]	Time 68.528 (75.531)	Loss 4874996736.0000 (4992640512.0000)	Prec@1 12.500 (11.719)   	Prec@5 53.906 (53.125)
INFO:root:
Test: [3/79]	Time 68.462 (73.175)	Loss 4973036032.0000 (4986105685.3333)	Prec@1 8.594 (10.677)   	Prec@5 52.344 (52.865)

**Comment by Dong-Jie-Chen** [comment]
Thanks for your reply! I've tried based on your suggestion, and it works!
